### ISSUE TITLE
Fish tag formats--cleanup and entry tool

### ISSUE THREAD

**Issue by kaijagahm** [MFEh2o/db#150]
We decided that it would be nice to have a function in the fish entry tool to check the format of all pit and floy tag numbers entered.

In order to do this, we have to have a clear idea of what formats are acceptable.

Note that it's _not_ sufficient to look at what the formats "should" be going forward, because we also have to account for old tags that get recaptured. All past formats are fair game, at least for the recaptured tags.

Mechanism for overriding the function in case a legitimate new tag format shows up--either or both of the following:
a) allow force, so that the person entering the fish data can allow those new tags in real time
b) (eventually) update the template entry tool itself, changing the regex to allow the new format.

This check is going to require writing regular expressions. People strongly dislike regex. I promise it's not so bad! Here are some resources that can help:
[Working with strings in R](https://evoldyn.gitlab.io/evomics-2018/ref-sheets/R_strings.pdf)
[Regular expressions in R](https://www.marcusrb.com/files/pdf/r/regex.pdf)
[Regex tester](https://regex101.com/)--you *definitely* want to test your regular expressions as you're writing them to make sure they actually do what you think they do.

Before I leave, I'll write the skeleton of a tag format check in the entry tool, to be filled in with regular expressions once the tag formats have been cleaned up/decided on.

BUT: before this check can be built in the fish entry tool, the existing tag formats in the database need to be cleaned up. We currently have a whole lot of different formats, and probably many of them are mistakes or shorthands. I already did some work on this in #22, but it's not 100% cleaned up.

--------- Preliminary exploration of existing tag formats -----------
Finding all unique tag formats is actually [a pretty interesting problem](https://twitter.com/kaija_bean/status/1400931182569734147). To tackle it, I wrote a "reverse regex" function:

```
# Load packages
library(checkmate) # for checks
library(dplyr) # for pipe

# The function takes a character vector, `vec`, as input. By default (unique = F), it returns one 'general' tag format for each element of the input vector. Alternatively, if you specify unique = T, it just returns a vector of all the unique formats.
# Basically what we're doing here is converting all the letters and numbers in each tag to placeholder values: 0 for digits, A for uppercase letters, and a for lowercase letters. We preserve all other characters (punctuation, spaces, etc) as they are, since those make for different formats.

rr <- function(vec, unique = F){
  # Check that `vec` is a character vector that has at least some non-missing entries
  checkmate::assertCharacter(vec, all.missing = F)
  
  # Classify each character as uppercase (A), lowercase (a), or digit (0). Leave punctuation marks as they are (because for my use case, different punctuation = different format)
  classified <- vec %>%
    stringr::str_replace_all(., "[0-9]", "0") %>%
    stringr::str_replace_all(., "[A-Z]", "A") %>%
    stringr::str_replace_all(., "[a-z]", "a")
  
  # Two possible returns: all unique formats, or all formats
  if(unique == T){
    # Get all unique formats
    unique_formats <- unique(classified)
    
    # Return unique formats
    return(unique_formats)
  }else{
    return(classified)
  }
}
```

Here's an example of using that function:
```
# Make an example vector
vec <- c("456AAA", "325ATC", "291WMG", "A.2134", "B.2058", "23_ABC")

# Run rr() with unique = F (default)
> rr(vec)
[1] "000AAA" "000AAA" "000AAA" "A.0000" "A.0000" "00_AAA"

# Run rr() with unique = T
> rr(vec, unique = T)
[1] "000AAA" "A.0000" "00_AAA"
```

Okay, so we have a way to reduce tags down to their unique formats. I ran this function on the pit and floy tags from the database, using this code:

```
library(tidyverse)
# source dbUtil.R, load other packages, connect to the database as usual.

fi <- dbTable("fish_info")

# Get all pit and floy tags
pits <- unique(c(fi$pitApply, fi$pitRecapture))
floys <- unique(c(fi$floyApply, fi$floyRecapture))

# Get unique formats for each
pitFormatsUnique <- rr(pits, unique = T)
pitFormatsUnique
floyFormatsUnique <- rr(floys, unique = T)
floyFormatsUnique

# Is there overlap between pit and floy formats? (yes)
pitFormatsUnique[pitFormatsUnique %in% floyFormatsUnique]
floyFormatsUnique[floyFormatsUnique %in% pitFormatsUnique]

# Show a simplified data frame with formats assigned to each tag
formats <- fi %>%
  # pivot to long format to make things easier to read
  select(contains("pit"), contains("floy")) %>%
  pivot_longer(cols = c("pitApply", "pitRecapture", "floyApply", "floyRecapture"), names_to = "type") %>%
  filter(!is.na(value)) %>%
  mutate(type = str_remove(type, "Recapture"),
         type = str_remove(type, "Apply")) %>%
  distinct() %>%
  mutate(format = rr(value))

head(formats)
```

We can see from this that we have quite a few different formats happening, and there's some overlap in formats between pit and floy (noted: we can't necessarily tell from the format of a tag number which type of tag it is!)

For an illustrative example, let's pare down the `formats` data frame to show the first occurrence of each unique format.
```
examplesPit <- formats %>%
   filter(type == "pit") %>%
   group_by(format) %>%
   slice(1)

examplesFloy <- formats %>%
   filter(type == "floy") %>%
   group_by(format) %>%
   slice(1)
```

**Comment by kaijagahm** [comment]
I've now written skeleton functions for pit and floy tag formats. I'm commenting out the code for those checks in fishEntry.R; you'll need to un-comment it.

For thoughts on how to match multiple regex strings, see these resources:
[Multiple strings with str_detect](https://stackoverflow.com/questions/35962426/multiple-strings-with-str-detect-r)
[Multiple patterns with filter and str_detect](https://stackoverflow.com/questions/44759180/filter-by-multiple-patterns-with-filter-and-str-detect)

**Comment by kaijagahm** [comment]
The fish entry tool is ready to have this check added. I've written almost the entire check function, in 'checks.R'. As of 2021-07-19, it's on lines 677-716, and I've reproduced the code here:

```
# checkTagFormats ---------------------------------------------------------
checkTagFormats <- function(new, fp = force_pitFormat, ff = force_floyFormat){
  # Check inputs
  assertDataFrame(new, col.names = "unique")
  assertSubset(c("pitApply", "pitRecapture", "floyApply", "floyRecapture"),
               choices = names(new))
  assertFlag(fp)
  assertFlag(ff)
  
  # Check tag formats
  ## Pit
  pitPattern <- "" # XXX NEED TO WRITE THIS! See GH issue #150
  pitProblems <- new %>%
    select(fishID, "apply" = pitApply, "recapture" = pitRecapture) %>%
    pivot_longer(cols = c("apply", "recapture"), names_to = "type", values_to = "tag") %>%
    filter(!is.na(tag),
           !str_detect(tag, pattern = pitPattern))
  
  if(nrow(pitProblems) > 0){
    stop(paste0("Some pit tag numbers don't match any expected pit tag formats:\n\n",
                paste0(capture.output(pitProblems), collapse = "\n"), 
                "\n\nIf you're sure you want to enter these tags, use ", 
                deparse(substitute(fp)), "."))
  }
  
  ## Floy
  floyPattern <- "" # XXX NEED TO WRITE THIS! See GH issue #150
  floyProblems <- new %>%
    select(fishID, "apply" = floyApply, "recapture" = floyRecapture) %>%
    pivot_longer(cols = c("apply", "recapture"), names_to = "type", values_to = "tag") %>%
    filter(!is.na(tag),
           !str_detect(tag, pattern = floyPattern))
  
  if(nrow(floyProblems) > 0){
    stop(paste0("Some floy tag numbers don't match any expected floy tag formats:\n\n",
                paste0(capture.output(floyProblems), collapse = "\n"), 
                "\n\nIf you're sure you want to enter these tags, use ", 
                deparse(substitute(ff)), "."))
  }
}
```

The variables `floyPattern` and `pitPattern` need to be added in as regex expressions. 

Then, once the function is finished, you'll need to add that check into 'fishEntry.R'. Currently, it's on line 376 and needs to be un-commented.

Of course, make sure to do these changes in the TEMPLATE repository. Preferably, do them after the end of the 2021 season and before the start of the 2022 season so that the 2022 repositories can be cloned from the template and incorporate these changes. Otherwise, you'll have to manually copy the change to the current-season repo as well as the template.

**Comment by ctsolomon** [comment]
Note that I have done a lot of work sorting through Long tag data that may be relevant to resolving this issue. My code on this is currently in 'canonical tags.R' on my hard drive.
